### PR TITLE
Fix bgw_launcher restart when failing during launcher setup

### PR DESCRIPTION
**Pull request by JLockerman**
We do cleanup for bgw_message_queue in the before_shmem_exit function,
but we were registering said function after initializing the database
hashtable. When bgw_launcher would crash in middle of populating the
hashtable, this would cause issues as the bgw_message_queue wouldn't be
reset, and the launcher would fail to restart as it would think that
another launcher was running. This commit moves the hook registering to
before we touch any of the state that the hook cleans up, in order to
ensure that the state will be cleaned regardless of where the failure
occurs. This is already safe for the bgw_message_queue since that
function validates that the caller owns the queue before altering it,
and is made safe for the hashtable via an additional level of
indirection, so we can check if the hashtable exists.

This commit also cleans up some of the lock ordering around the
bgw_message_queue to prevent potential race conditions; specifically, if
we're doing a test-and-set, both the "test" and the "set" have to be
within the same lock-acquisition.

### CODE REVIEW

File: src/loader/bgw_message_queue.c [hunk 1]
```diff
@@ -121,25 +121,37 @@ static void
...
-	vq->reader_pid = InvalidPid;
+	if (vq->reader_pid == MyProcPid)
+	{
+		reset = true;
+		vq->reader_pid = InvalidPid;
+	}
 	SpinLockRelease(&vq->mutex);
+
+	if (!reset)
+		ereport(ERROR, (errmsg("cannot read if not reader for TimescaleDB background worker message queue")));
```
Comment (svenklemm) [inline src/loader/bgw_message_queue.c]: You were just moving this but this error message sounds weird

Comment (davidkohn88) [reply src/loader/bgw_message_queue.c]: perhaps "multiple TimescaleDB background worker launchers have been started when only one is allowed" hint "please report this bug on our github page"

Comment (JLockerman) [reply src/loader/bgw_message_queue.c]: @svenklemm if I'm moving it anyway might as well fix it. What do you think of
```C
"tried to reset TimescaleDB background worker message queue we do not own"
```

Comment (svenklemm) [reply src/loader/bgw_message_queue.c]: @JLockerman definitely an improvement over current text

**Comment by JLockerman** [comment]
hey, codecov is back!

File: src/loader/bgw_message_queue.c [hunk 2]
```diff
@@ -121,25 +121,40 @@ static void
...
+		reset = true;
+		vq->reader_pid = InvalidPid;
+	}
 	SpinLockRelease(&vq->mutex);
+
+	if (!reset)
+		ereport(ERROR, (
+						ERRCODE_INTERNAL_ERROR,
+						errmsg("multiple TimescaleDB background worker launchers have been started when only one is allowed"),
+						errhint("this is a bug, please report it on our github page")));
```
Comment (cevian) [inline src/loader/bgw_message_queue.c]: nit: Hints should be complete sentences and start with capital letter and end with period

File: src/loader/bgw_message_queue.c [hunk 3]
```diff
@@ -121,25 +121,40 @@ static void
...
-		SpinLockRelease(&vq->mutex);
 	}
-	else if (queue_get_reader(queue) != MyProcPid)
-		ereport(ERROR, (errmsg("only one reader allowed for TimescaleDB background worker message queue")));
+	reader_pid = vq->reader_pid;
+	SpinLockRelease(&vq->mutex);
+	if (reader_pid != MyProcPid)
+		ereport(ERROR,
+				(errmsg("only one reader allowed for TimescaleDB background worker message queue"),
+				 errhint("Current process is %d", reader_pid)));
```
Comment (cevian) [inline src/loader/bgw_message_queue.c]: nit: should end with period

**Pull request merged**